server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java

```diff
@@ -52,97 +68,305 @@
-            }
-        }
+            waitForShardsReadyForClosing(index, state, timeout, new ActionListener<AcknowledgedResponse>() {
+                @Override
+                public void onResponse(final AcknowledgedResponse result) {
+                    results.put(index, result);
+                    processIfFinished();
+                }
 
-        if (indicesToClose.isEmpty()) {
-            return currentState;
+                @Override
+                public void onFailure(final Exception e) {
+                    results.put(index, new AcknowledgedResponse(false));
+                    processIfFinished();
+                }
+
+                private void processIfFinished() {
+                    if (countDown.countDown()) {
+                        finalizeClosing(unmodifiableMap(results), listener);
+                    }
+                }
+            });
         }
+    }
 
-        // Check if index closing conflicts with any running restores
-        RestoreService.checkIndexClosing(currentState, indicesToClose);
-        // Check if index closing conflicts with any running snapshots
-        SnapshotsService.checkIndexClosing(currentState, indicesToClose);
-        logger.info("closing indices [{}]", indicesAsString);
-
-        MetaData.Builder mdBuilder = MetaData.builder(currentState.metaData());
-        ClusterBlocks.Builder blocksBuilder = ClusterBlocks.builder()
-            .blocks(currentState.blocks());
-        for (IndexMetaData openIndexMetadata : indicesToClose) {
-            final String indexName = openIndexMetadata.getIndex().getName();
-            mdBuilder.put(IndexMetaData.builder(openIndexMetadata).state(IndexMetaData.State.CLOSE));
-            blocksBuilder.addIndexBlock(indexName, INDEX_CLOSED_BLOCK);
+    private void waitForShardsReadyForClosing(final Index index, final ClusterState state, @Nullable final TimeValue timeout,
+                                              final ActionListener<AcknowledgedResponse> listener) {
+        final IndexMetaData indexMetaData = state.metaData().index(index);
+        if (indexMetaData == null || indexMetaData.getState() == IndexMetaData.State.CLOSE) {
+            logger.debug("index {} has been blocked before closing and is already closed, ignoring", index);
+            listener.onResponse(new AcknowledgedResponse(true));
+            return;
         }
+        final IndexRoutingTable indexRoutingTable = state.routingTable().index(index);
+        if (indexRoutingTable == null) {
+            logger.debug("index {} has been blocked before closing but is now deleted, ignoring", index);
+            listener.onResponse(new AcknowledgedResponse(true));
+            return;
+        } else if (indexRoutingTable.allPrimaryShardsUnassigned()) {
+            logger.debug("index {} has been blocked before closing but is now unassigned, ignoring", index);
+            listener.onResponse(new AcknowledgedResponse(true));
```

Comment:
I wonder if we should do this check per IndexShardRoutingTable as well, otherwise we fail if some primary shards are unassigned while others are not